### PR TITLE
Refactor database schema

### PR DESCRIPTION
**Pull request by fmrsabino**
In order to simplify the schema that was set for registering accounts and notification channels (to which a user can subscribe to), the following was changed:

### DataSource

- The `emails` schema was removed – the added tables are no longer present under a schema which simplifies the initial setup of the database.
- The table `emails.subscriptions` was renamed to `notification_types` – this is more in line with the contents that the table stores (i.e.: types of notification that a user can subscribe to).
- The table `emails.account_subscriptions` was renamed to `subscriptions` – this table stores the subscriptions that each user has (to a specific `channel`).
- The `IEmailDataSource` was renamed to `IAccountDataSource`
  * `getEmail` was renamed to `getAccount`
  * `saveEmail` was renamed to `safeAccount`
  * `deleteEmail` was renamed to `deleteAccount`
  * `updateEmail` was renamed to `updateAccountEmail`

### Domain

- `EmailRepository` was renamed to `AccountRepository`
- The `Email` entity was renamed to `Account`.
- `EmailSaveError` was renamed to `AccountSaveError`

### Routes

- Renamed email payload keys from `account` to `signer`.

### CODE REVIEW

File: src/domain/alerts/alerts.repository.ts [hunk 1]
```diff
@@ -26,8 +26,8 @@ export class AlertsRepository implements IAlertsRepository {
     private readonly alertsApi: IAlertsApi,
     @Inject(IEmailApi)
     private readonly emailApi: IEmailApi,
-    @Inject(IEmailRepository)
-    private readonly emailRepository: IEmailRepository,
+    @Inject(IAccountRepository)
+    private readonly emailRepository: IAccountRepository,
```
Comment (hectorgomezv) [inline src/domain/alerts/alerts.repository.ts:30]: I'd rename this variable also to `accountRepository`

File: src/domain/interfaces/account.datasource.interface.ts [hunk 2]
```diff
@@ -22,11 +25,11 @@ export interface IEmailDataSource {
    * @param args.safeAddress - the Safe address to use as filter
    * @param args.account - the owner address to which link the email address is linked to
    */
-  getEmail(args: {
+  getAccount(args: {
```
Comment (hectorgomezv) [inline src/domain/interfaces/account.datasource.interface.ts:28]: I think this function documentation needs to be changed as well (from the concept `email` to the `account` one).

File: src/domain/interfaces/account.datasource.interface.ts [hunk 3]
```diff
@@ -38,7 +41,7 @@ export interface IEmailDataSource {
    * @param args.code - the generated code to be used to verify this email address
    * @param args.verificationGeneratedOn – the date which represents when the code was generated
    */
-  saveEmail(args: {
+  saveAccount(args: {
```
Comment (hectorgomezv) [inline src/domain/interfaces/account.datasource.interface.ts:44]: As above, I think this function documentation needs to be changed (from the concept `email` to the `account` one).

File: src/routes/email/exception-filters/email-does-not-exist.exception-filter.ts [hunk 4]
```diff
@@ -5,13 +5,13 @@ import {
   HttpStatus,
 } from '@nestjs/common';
 import { Response } from 'express';
-import { EmailAddressDoesNotExistError } from '@/datasources/email/errors/email-address-does-not-exist.error';
+import { AccountDoesNotExistError } from '@/datasources/account/errors/account-does-not-exist.error';
 
-@Catch(EmailAddressDoesNotExistError)
+@Catch(AccountDoesNotExistError)
 export class EmailAddressDoesNotExistExceptionFilter
```
Comment (hectorgomezv) [inline src/routes/email/exception-filters/email-does-not-exist.exception-filter.ts:11]: I think this class also needs to be renamed, right? (and its filename)

File: src/datasources/account/__tests__/test.account.datasource.module.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+const accountDataSource = {
+  getVerifiedAccountEmailsBySafeAddress: jest.fn(),
+  getAccount: jest.fn(),
+  saveAccount: jest.fn(),
+  setVerificationCode: jest.fn(),
+  setVerificationSentDate: jest.fn(),
+  verifyEmail: jest.fn(),
+  deleteAccount: jest.fn(),
+  updateAccountEmail: jest.fn(),
+} as unknown as IAccountDataSource;
```
Comment (iamacook) [inline src/datasources/account/__tests__/test.account.datasource.module.ts:13]:
```suggestion
} as jest.MockedObjectDeep<IAccountDataSource>;
```

File: src/datasources/account/account.datasource.ts [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+      isVerified: account.verified,
+      safeAddress: account.safe_address,
+      account: account.account,
+      verificationCode: account.verification_code,
+      verificationGeneratedOn: account.verification_code_generated_on,
+      verificationSentOn: account.verification_sent_on,
+    };
+  }
+
+  async saveAccount(args: {
```
Comment (iamacook) [inline src/datasources/account/account.datasource.ts:79]: Nit: what do you think about renaming (and propagating) this to `createAccount` as this is inserting, not updating.

File: src/domain/account/account.repository.interface.ts [hunk 7]
```diff
@@ -13,14 +13,14 @@ export interface IEmailRepository {
   }): Promise<string[]>;
 
   /**
-   * Saves an email entry.
+   * Saves an account entry.
    *
    * @param args.chainId - the chain id of where the Safe is deployed
    * @param args.safeAddress - the Safe address to which we should store the email address
    * @param args.emailAddress - the email address to store
    * @param args.account - the owner address to which we should link the email address to
```
Comment (iamacook) [inline src/domain/account/account.repository.interface.ts:21]: If we agree on renaming this to `createAccount` then we should update this.

File: src/domain/account/errors/account-save.error.ts [hunk 8]
```diff
@@ -1,4 +1,4 @@
-export class EmailSaveError extends Error {
+export class AccountSaveError extends Error {
   constructor(chainId: string, safeAddress: string, account: string) {
     super(
       `Error while saving provided email was not saved. chainId=${chainId}, safeAddress=${safeAddress}, account=${account}`,
```
Comment (iamacook) [inline src/domain/account/errors/account-save.error.ts:4]: Here too if we change to `createAccount`.

File: src/datasources/account/account.datasource.module.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+import { AccountDataSource } from '@/datasources/account/account.datasource';
+import { IAccountDataSource } from '@/domain/interfaces/account.datasource.interface';
+import { PostgresDatabaseModule } from '@/datasources/db/postgres-database.module';
+
+@Module({
+  imports: [PostgresDatabaseModule],
+  providers: [{ provide: IAccountDataSource, useClass: AccountDataSource }],
+  exports: [IAccountDataSource],
+})
+export class AccountDatasourceModule {}
```
[the remark this answers is not in the document]
Comment (fmrsabino) [reply src/datasources/account/account.datasource.module.ts]: [fa63a4e](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/fa63a4e9950036c1b044c936774359b805c7c62f)

File: src/domain/account/account.repository.interface.ts [hunk 10]
```diff
@@ -64,13 +64,13 @@ export interface IEmailRepository {
   }): Promise<any>;
 
   /**
-   * Deletes an email address for a given account.
+   * Deletes an account.
    *
    * @param args.chainId - the chain id of where the Safe is deployed
    * @param args.safeAddress - the Safe address to which we should remove the email address from
```
[the remark this answers is not in the document]
Comment (fmrsabino) [reply src/domain/account/account.repository.interface.ts]: [fa63a4e](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/fa63a4e9950036c1b044c936774359b805c7c62f)

File: src/domain/account/account.repository.interface.ts [hunk 11]
```diff
@@ -64,13 +64,13 @@ export interface IEmailRepository {
...
    * @param args.account - the owner address to which we should remove the email address from
```
[the remark this answers is not in the document]
Comment (fmrsabino) [reply src/domain/account/account.repository.interface.ts]: [fa63a4e](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/fa63a4e9950036c1b044c936774359b805c7c62f)

File: src/domain/alerts/alerts.repository.ts (hunk 1 above)
Comment (fmrsabino) [reply src/domain/alerts/alerts.repository.ts]: [fa63a4e](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/fa63a4e9950036c1b044c936774359b805c7c62f)

File: src/domain/interfaces/account.datasource.interface.ts (hunk 2 above)
Comment (fmrsabino) [reply src/domain/interfaces/account.datasource.interface.ts:28]: [fa63a4e](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/fa63a4e9950036c1b044c936774359b805c7c62f)

File: src/domain/interfaces/account.datasource.interface.ts (hunk 3 above)
Comment (fmrsabino) [reply src/domain/interfaces/account.datasource.interface.ts]: [fa63a4e](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/fa63a4e9950036c1b044c936774359b805c7c62f)

File: src/routes/email/exception-filters/email-does-not-exist.exception-filter.ts (hunk 4 above)
Comment (fmrsabino) [reply src/routes/email/exception-filters/email-does-not-exist.exception-filter.ts]: [fa63a4e](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/fa63a4e9950036c1b044c936774359b805c7c62f)

File: src/datasources/email/account.datasource.ts [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+      isVerified: account.verified,
+      safeAddress: account.safe_address,
+      account: account.account,
+      verificationCode: account.verification_code,
+      verificationGeneratedOn: account.verification_code_generated_on,
+      verificationSentOn: account.verification_sent_on,
+    };
+  }
+
+  async saveAccount(args: {
```
[the remark this answers is not in the document]
Comment (fmrsabino) [reply src/datasources/email/account.datasource.ts]: [fa63a4e](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/fa63a4e9950036c1b044c936774359b805c7c62f)

File: src/datasources/email/account.datasource.ts [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+    const subscriptions = await this.sql`
+        WITH deleted_subscriptions AS (
+            DELETE FROM subscriptions
+                WHERE account_id = (SELECT id
+                                    FROM accounts
+                                    WHERE unsubscription_token = ${args.token})
+                RETURNING notification_type)
+        SELECT subs.key, subs.name
+        FROM notification_types subs
+                 JOIN deleted_subscriptions deleted_subs ON subs.id = deleted_subs.notification_type;;
```
[the remark this answers is not in the document]
Comment (fmrsabino) [reply src/datasources/email/account.datasource.ts]: [fa63a4e](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/fa63a4e9950036c1b044c936774359b805c7c62f)

File: src/datasources/email/errors/account-does-not-exist.error.ts [hunk 14]
```diff
@@ -1,4 +1,4 @@
-export class EmailAddressDoesNotExistError extends Error {
+export class AccountDoesNotExistError extends Error {
```
[the remark this answers is not in the document]
Comment (fmrsabino) [reply src/datasources/email/errors/account-does-not-exist.error.ts]: [fa63a4e](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/fa63a4e9950036c1b044c936774359b805c7c62f)

File: src/datasources/account/account.datasource.ts (hunk 6 above)
Comment (fmrsabino) [reply src/datasources/account/account.datasource.ts]: [fa63a4e](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/fa63a4e9950036c1b044c936774359b805c7c62f)

File: src/domain/account/account.repository.interface.ts (hunk 7 above)
Comment (fmrsabino) [reply src/domain/account/account.repository.interface.ts]: [fa63a4e](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/fa63a4e9950036c1b044c936774359b805c7c62f)

File: src/domain/account/errors/account-save.error.ts (hunk 8 above)
Comment (fmrsabino) [reply src/domain/account/errors/account-save.error.ts:4]: [fa63a4e](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/fa63a4e9950036c1b044c936774359b805c7c62f)

File: src/routes/email/email.controller.module.ts [hunk 15]
```diff
@@ -1,10 +1,10 @@
 import { Module } from '@nestjs/common';
-import { EmailDomainModule } from '@/domain/email/email.domain.module';
+import { AccountDomainModule } from '@/domain/account/account.domain.module';
 import { EmailController } from '@/routes/email/email.controller';
 import { EmailService } from '@/routes/email/email.service';
 
 @Module({
-  imports: [EmailDomainModule],
+  imports: [AccountDomainModule],
```
[the remark this answers is not in the document]
Comment (fmrsabino) [reply src/routes/email/email.controller.module.ts:7]: The goal of this PR was not to change the routes (at least not yet). The resource path for these routes are still under `/email`.

File: src/datasources/email/account.datasource.ts [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+    if (!account) {
+      throw new AccountDoesNotExistError(
+        args.chainId,
+        args.safeAddress,
+        args.account,
+      );
+    }
+  }
+
+  async setVerificationSentDate(args: {
```
[the remark this answers is not in the document]
Comment (fmrsabino) [reply src/datasources/email/account.datasource.ts]: [069a023](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/069a0230a184c3456adb4e4b97f4811f598931e2)

File: src/datasources/account/account.datasource.ts [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+}
+
+interface Subscription {
+  id: number;
+  key: string;
+  name: string;
+}
+
+@Injectable()
+export class AccountDataSource implements IAccountDataSource {
```
Comment (iamacook) [inline src/datasources/account/account.datasource.ts:31]: Something to consider (in a separate PR) is ensuring all of the saved addresses are checksummed.

File: src/domain/account/account.repository.interface.ts [hunk 18]
```diff
@@ -13,26 +13,26 @@ export interface IEmailRepository {
...
 
   /**
-   * Saves an email entry.
+   * Creates a new account.
    *
    * @param args.chainId - the chain id of where the Safe is deployed
    * @param args.safeAddress - the Safe address to which we should store the email address
    * @param args.emailAddress - the email address to store
-   * @param args.account - the owner address to which we should link the email address to
+   * @param args.signer - the owner address to which we should link the email address to
```
Comment (iamacook) [inline src/domain/account/account.repository.interface.ts:21]:
```suggestion
   * @param args.safeAddress - the Safe address which we should create the account for
   * @param args.emailAddress - the email address to store
   * @param args.signer - the owner address to which we should link the account to
```

File: src/domain/account/errors/account-save.error.ts (hunk 8 above)
Comment (iamacook) [reply src/domain/account/errors/account-save.error.ts:4]: The message needs to be tweaked a bit:

```suggestion
      `Error while creating account. Account was not created. chainId=${chainId}, safeAddress=${safeAddress}, signer=${signer}`,
```

File: src/domain/account/errors/account-save.error.ts [hunk 19]
```diff
@@ -1,4 +1,4 @@
-export class EmailSaveError extends Error {
+export class AccountSaveError extends Error {
   constructor(chainId: string, safeAddress: string, account: string) {
```
Comment (iamacook) [inline src/domain/account/errors/account-save.error.ts:2]:
```suggestion
  constructor(chainId: string, safeAddress: string, signer: string) {
```

File: src/domain/account/errors/email-already-verified.error.ts [hunk 20]
```diff
@@ -0,0 +1,10 @@
+export class EmailAlreadyVerifiedError extends Error {
+  readonly signer: string;
+
+  constructor(args: { chainId: string; safeAddress: string; signer: string }) {
+    super(
+      `The email address is already verified. chainId=${args.chainId}, safeAddress=${args.safeAddress}, signer=${args.signer}`,
+    );
+    this.signer = args.signer;
```
Comment (iamacook) [inline src/domain/account/errors/email-already-verified.error.ts:8]:
```suggestion
```

File: src/domain/account/errors/verification-timeframe.error.ts [hunk 21]
```diff
@@ -2,12 +2,12 @@ export class ResendVerificationTimespanError extends Error {
...
     chainId: string;
     safeAddress: string;
-    account: string;
+    signer: string;
     timespanMs: number;
     lockWindowMs: number;
   }) {
     super(
-      `Verification cannot be re-sent at this time. ${args.timespanMs} ms have elapsed out of ${args.lockWindowMs} ms for account=${args.account}, safe=${args.safeAddress}, chainId=${args.chainId}`,
+      `Verification cannot be re-sent at this time. ${args.timespanMs} ms have elapsed out of ${args.lockWindowMs} ms for signer=${args.signer}, safe=${args.safeAddress}, chainId=${args.chainId}`,
```
Comment (iamacook) [inline src/domain/account/errors/verification-timeframe.error.ts:10]:
```suggestion
      `Verification cannot be resent at this time. ${args.timespanMs} ms have elapsed out of ${args.lockWindowMs} ms for signer=${args.signer}, safe=${args.safeAddress}, chainId=${args.chainId}`,
```

File: src/domain/interfaces/account.datasource.interface.ts [hunk 22]
```diff
@@ -71,114 +74,113 @@ export interface IEmailDataSource {
...
    */
   verifyEmail(args: {
     chainId: string;
     safeAddress: string;
-    account: string;
+    signer: string;
   }): Promise<void>;
 
   /**
    * Deletes an email address for a given account.
```
Comment (iamacook) [inline src/domain/interfaces/account.datasource.interface.ts:102]:
```suggestion
   * Deletes the given account.
```

File: src/routes/email/email.controller.delete-email.spec.ts [hunk 23]
```diff
@@ -74,48 +74,48 @@ describe('Email controller delete email tests', () => {
...
+    const signature = await signer.signMessage({ message });
+    accountDataSource.getAccount.mockResolvedValue(account);
+    accountDataSource.deleteAccount.mockImplementation(() => Promise.resolve());
     emailApi.deleteEmailAddress.mockResolvedValue();
 
     await request(app.getHttpServer())
       .delete(`/v1/chains/${chain.chainId}/safes/${safeAddress}/emails`)
       .send({
-        account: account.address,
+        account: signer.address,
```
Comment (iamacook) [inline src/routes/email/email.controller.delete-email.spec.ts:94]: We should also change this to `signer` if we are doing so elsewhere.

File: src/routes/email/email.controller.delete-email.spec.ts [hunk 24]
```diff
@@ -242,20 +238,20 @@ describe('Email controller delete email tests', () => {
...
     });
-    emailDatasource.getEmail.mockResolvedValueOnce(email);
+    accountDataSource.getAccount.mockResolvedValueOnce(account);
     emailApi.deleteEmailAddress.mockRejectedValue(new Error('Some error'));
 
     await request(app.getHttpServer())
       .delete(`/v1/chains/${chain.chainId}/safes/${safe.address}/emails`)
       .send({
-        account: account.address,
+        account: signer.address,
```
Comment (iamacook) [inline src/routes/email/email.controller.delete-email.spec.ts:247]: On the email routes, `account` is used. I'm impartial as to whether we keep them as `account` but wanted to point out that they are no longer the same as what is mentioned in errors (`signer`).

File: src/domain/account/errors/email-already-verified.error.ts (hunk 20 above)
Comment (fmrsabino) [reply src/domain/account/errors/email-already-verified.error.ts:8]: The property is used in the `EmailAlreadyVerifiedExceptionFilter` so we need to store it or?

File: src/routes/email/email.controller.delete-email.spec.ts (hunk 24 above)
Comment (fmrsabino) [reply src/routes/email/email.controller.delete-email.spec.ts:247]: I agree. Will update the API spec accordingly.

Changed in [d56dd44](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/d56dd4443b128abb5be9c79161c5d9b8bcc4b8ed)

File: src/routes/email/email.controller.delete-email.spec.ts (hunk 23 above)
Comment (fmrsabino) [reply src/routes/email/email.controller.delete-email.spec.ts:94]: [d56dd44](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/d56dd4443b128abb5be9c79161c5d9b8bcc4b8ed)

File: src/domain/account/errors/verification-timeframe.error.ts (hunk 21 above)
Comment (fmrsabino) [reply src/domain/account/errors/verification-timeframe.error.ts]: [d56dd44](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/d56dd4443b128abb5be9c79161c5d9b8bcc4b8ed)

File: src/domain/account/errors/account-save.error.ts (hunk 19 above)
Comment (fmrsabino) [reply src/domain/account/errors/account-save.error.ts]: [d56dd44](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/d56dd4443b128abb5be9c79161c5d9b8bcc4b8ed)

File: src/domain/account/account.repository.interface.ts (hunk 18 above)
Comment (fmrsabino) [reply src/domain/account/account.repository.interface.ts]: [d56dd44](https://github.com/safe-global/safe-client-gateway/pull/1050/commits/d56dd4443b128abb5be9c79161c5d9b8bcc4b8ed)

File: src/domain/account/errors/email-already-verified.error.ts (hunk 20 above)
Comment (iamacook) [reply src/domain/account/errors/email-already-verified.error.ts:8]: Indeed 🦅

**Review by iamacook — APPROVED** [review]

**Review by hectorgomezv — APPROVED** [review]
🚀

**Pull request merged**